cmd/thanos/remote-write-receive.go

```diff
@@ -0,0 +1,281 @@
+package main
+
+import (
+	"context"
+	"fmt"
+	"math"
+	"net"
+	"sync"
+	"time"
+
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+	"github.com/improbable-eng/thanos/pkg/cluster"
+	"github.com/improbable-eng/thanos/pkg/receive"
+	"github.com/improbable-eng/thanos/pkg/runutil"
+	"github.com/improbable-eng/thanos/pkg/store"
+	"github.com/improbable-eng/thanos/pkg/store/storepb"
+	"github.com/oklog/run"
+	opentracing "github.com/opentracing/opentracing-go"
+	"github.com/pkg/errors"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/common/model"
+	"github.com/prometheus/prometheus/storage/tsdb"
+	"google.golang.org/grpc"
+	kingpin "gopkg.in/alecthomas/kingpin.v2"
+)
+
+func registerRemoteWriteReceive(m map[string]setupFunc, app *kingpin.Application, name string) {
+	cmd := app.Command(name, "accept Prometheus remote write API requests (EXPERIMENTAL, this may change drastically without notice)")
+
+	grpcBindAddr, httpMetricsBindAddr, cert, key, clientCA, newPeerFn := regCommonServerFlags(cmd)
+
+	remoteWriteAddress := cmd.Flag("remote-write.address", "Address to listen on for remote write requests.").
+		Default("0.0.0.0:19291").String()
+
+	dataDir := cmd.Flag("tsdb.path", "Data directory of TSDB.").
+		Default("./data").String()
+
+	m[name] = func(g *run.Group, logger log.Logger, reg *prometheus.Registry, tracer opentracing.Tracer, _ bool) error {
+		peer, err := newPeerFn(logger, reg, false, "", false)
+		if err != nil {
+			return errors.Wrap(err, "new cluster peer")
+		}
+		return runReceiver(
+			g,
+			logger,
+			reg,
+			tracer,
+			*grpcBindAddr,
+			*cert,
+			*key,
+			*clientCA,
+			*httpMetricsBindAddr,
+			*remoteWriteAddress,
+			*dataDir,
+			peer,
+			name,
+		)
+	}
+}
+
+func runReceiver(
+	g *run.Group,
+	logger log.Logger,
+	reg *prometheus.Registry,
+	tracer opentracing.Tracer,
+	grpcBindAddr string,
+	cert string,
+	key string,
+	clientCA string,
+	httpMetricsBindAddr string,
+	remoteWriteAddress string,
+	dataDir string,
+	peer cluster.Peer,
+	component string,
+) error {
+	level.Info(logger).Log("msg", "setting up receiver")
+
+	var metadata = &promMetadata{
+		// Start out with the full time range. The shipper will constrain it later.
+		// TODO(fabxc): minimum timestamp is never adjusted if shipping is disabled.
```

Comment:
Was this code copied from somewhere? Slightly confusing to see `TODO(fabxc)` in here, but not a big deal I guess.

Reply:
haha yea was surprised as well at  the beginning (: It is copied indeed.